src/core/indexer/mod.rs

```diff
@@ -0,0 +1,595 @@
+                "Need '{l}' more characters",
+                l = ((cs as usize) - qb64.len()),
+            )));
+        }
+
+        let index = util::b64_to_u32(&qb64[hs..(hs + ms as usize)])?;
+        let odx = &qb64[(hs + ms as usize)..(hs + ms as usize + szg.os as usize)];
+
+        let mut ondex: Option<u32> = None;
+        if CurrentSigCodex::has_code(hard) {
+            if szg.os != 0 {
+                ondex = Some(util::b64_to_u32(odx)?);
+            }
+            // not zero or None
+            if ondex.is_some() && ondex.unwrap() != 0 {
+                return err!(Error::Value(format!(
+                    "Invalid ondex = '{o}' for code = '{hard}'.",
+                    o = ondex.unwrap()
+                )));
+            }
+        } else if szg.os != 0 {
+            ondex = Some(util::b64_to_u32(odx)?);
+        } else {
+            ondex = Some(index);
+        }
+
+        // index is index for some codes and variable length for others
+        let mut fs = szg.fs;
+        if fs == 0 {
+            if (cs % 4) != 0 {
+                return err!(Error::Validation(format!(
+                    "Whole code size not multiple of 4 for variable length material. cs = '{cs}'"
+                )));
+            }
+
+            if szg.os != 0 {
+                return err!(Error::Validation(format!(
+                    "Non-zero other index size for variable length material. os = '{o}'",
+                    o = szg.os
+                )));
+            }
+
+            fs = (index * 4) + cs;
+        }
+
+        if qb64.len() < (fs as usize) {
+            return err!(Error::Shortage(format!(
+                "Need '{m}' more chars.",
+                m = { fs as usize - qb64.len() }
+            )));
+        }
+
+        let qb64_ = &qb64[..fs as usize];
+        let ps = cs % 4;
+        let pbs = 2 * if ps != 0 { ps } else { szg.ls };
+
+        let raw: Vec<u8>;
+        if ps != 0 {
+            let mut buf = "A".repeat(ps as usize);
+            buf.push_str(&qb64_[(cs as usize)..]);
+
+            let mut paw = Vec::<u8>::new();
+            base64::engine::general_purpose::URL_SAFE.decode_vec(buf, &mut paw)?;
+
+            let mut pi: i32 = 0;
+            for b in &paw[..ps as usize] {
+                pi = (pi * 256) + (*b as i32)
+            }
+
+            if (pi & (2_i32.pow(pbs) - 1)) != 0 {
+                return err!(Error::Prepad());
+            }
+
+            raw = paw[ps as usize..].to_owned();
+            paw.clear();
+        } else {
+            let buf = &qb64_[cs as usize..];
+            let mut paw = Vec::<u8>::new();
+            base64::engine::general_purpose::URL_SAFE.decode_vec(buf, &mut paw)?;
+
+            let mut li: u32 = 0;
+            for b in &paw[..szg.ls as usize] {
+                li = (li * 256) + (*b as u32);
+            }
+
+            if li != 0 {
+                return if szg.ls == 1 {
+                    err!(Error::NonZeroedLeadByte())
```

Comment:
I ported this code in one of the other classes too, but it feels a little specific to have two errors for these two cases. I guess we can address that in the error cleanup issue.

Reply:
We could pass a param instead, of the number of bytes - for example

Reply:
added to the error issue #51